corehq/blobs/tasks.py

```diff
@@ -14,19 +14,25 @@ def delete_expired_blobs():
     blob_expirations = BlobExpiration.objects.filter(expires_on__lt=_utcnow(), deleted=False)
 
     db = get_blob_db()
-    paths = []
-    bytes_deleted = 0
-    for blob_expiration in blob_expirations:
-        paths.append(db.get_path(blob_expiration.identifier, blob_expiration.bucket))
-        bytes_deleted += blob_expiration.length
-
-    db.bulk_delete(paths)
-    blob_expirations.update(deleted=True)
-    datadog_counter(
-        'commcare.temp_blobs.bytes_deleted',
-        value=bytes_deleted,
-    )
-    return bytes_deleted
+    total_bytes_deleted = 0
+    while blob_expirations.exists():
+        paths = []
+        deleted_ids = []
+        bytes_deleted = 0
+        for blob_expiration in blob_expirations[:1000]:
+            paths.append(db.get_path(blob_expiration.identifier, blob_expiration.bucket))
+            deleted_ids.append(blob_expiration.id)
+            bytes_deleted += blob_expiration.length
+
+        db.bulk_delete(paths)
+        BlobExpiration.objects.filter(id__in=deleted_ids).update(deleted=True)
+        datadog_counter(
+            'commcare.temp_blobs.bytes_deleted',
+            value=bytes_deleted,
+        )
+        total_bytes_deleted += bytes_deleted
+
+    return total_bytes_deleted
```

Comment:
would it be bad if two of these tasks are executing simultaneously? that might happen if this takes longer than 12 hours to run. in which case using locks with a large timeout could be used to ensure only one is running at a time

Reply:
is this return value actually used for anything other than tests?

Reply:
i don't know, it might be used if something is inspecting the celery result.

i was asking more because i don't know if having two processes trying to delete the same blobs might be a problem

Reply:
I don't think anything uses the return value since this task is never imported anywhere.

I also don't think it matters if multiple instances of this task are running simultaneously, although the stats reported to datadog could be incorrect if that happened. I'm going to re-code it so it spawns a new task after deleting each batch. That way a single task should never run for very long (although there still could be overlap between the subsequently spawned tasks and the periodic task).

Reply:
3a41273